prov/efa/src/rdm/efa_rdm_srx.c

```diff
@@ -42,12 +87,46 @@
  * @param[in] addr the source address of the incoming message
  * @param[in] size the size of the incoming message
  * @param[out] peer_rx_entry the obtained peer_rx_entry
- * @return int 0 on success, a negative integer on failure
+ * @return int FI_SUCCESS when a matched rx entry is found, -FI_ENOENT when the
+ * match is not found, other negative integer for other errors.
  */
 static int efa_rdm_srx_get_msg(struct fid_peer_srx *srx, fi_addr_t addr,
 		       size_t size, struct fi_peer_rx_entry **peer_rx_entry)
 {
-    return -FI_ENOSYS;
+	struct rxr_ep *rxr_ep;
+	struct rxr_pkt_entry *rx_pkt_entry;
+	struct rxr_op_entry *rx_entry;
+	int ret;
+
+	rxr_ep = (struct rxr_ep *) srx->ep_fid.fid.context;
+	/*
+	 * TODO:
+	 * In theory we should not need to allocate a pkt_entry to get a rx entry,
+	 * but the current EFA code needs an pkt_entry as input. A major refactor
+	 * is needed to split the context (addr, size etc.) and data from pkt entry
+	 * and make rxr_*_get_*_rx_entry needs context only.
+	 */
+	rx_pkt_entry = rxr_pkt_entry_alloc(rxr_ep, rxr_ep->efa_rx_pkt_pool, RXR_PKT_FROM_EFA_RX_POOL);
```

Comment:
Should we be concerned about memory exhaustion? `get_msg` will get called for every message received by either the SHM or EFA endpoints, so we are effectively using two `pkt_entry`s for every message. I know we release this `pkt_entry` at the end of this function.

Reply:
We can have a separate pkt pool for it, like shm rx pkt pool. But we may need to rename it as peer srx pkt pool or something similar. I was hesitant to do it within this PR as I hope to make this pkt pool change in the PR that integrates with shm.

Reply:
> get_msg will get called for every message received by either the SHM or EFA endpoints, so we are effectively using two pkt_entrys for every message. so we are effectively using two pkt_entrys for every message.

This is not true, `get_msg` will only be called by shm provider.

Reply:
I think I can make it allocated from shm_rx_pool in this PR, but add a TODO to change it. Does that sound OK to you?

Reply:
> This is not true, `get_msg` will only be called by shm provider.

You're right. `start_msg` is called by both EFA and SHM providers but `get_msg` is only called by SHM

> I think I can make it allocated from shm_rx_pool in this PR, but add a TODO to change it. Does that sound OK to you?

I don't have enough experience to say for sure. On instances with high core count, shm pool will be used heavily. Could we make the pools larger?

Reply:
This PR will not change any behavior for the current code. In the integration, we will need to make `get_msg` allocate pkt entry from a dedicated pool. 

> On instances with high core count, shm pool will be used heavily. Could we make the pools larger?
I do not think we will use more memory after the integration. This pkt pool is basically a replacement of the current `shm_rx_pool`, and we will drop shm_tx_pool and shm_rx_pool. The memory usage is actually smaller.

Reply:
My concern wasn't the total memory usage but the usage within each pool.

All `efa_*_pool` and `shm_*_pool` are fixed size and cannot grow. I'll let you make the best decision.

https://github.com/ofiwg/libfabric/blob/8bb6f708e175ed9a8845142b715340ccdab75f79/prov/efa/src/rdm/rxr_ep.c#L1137-L1245

Reply:
Ah I get your point. Currently efa will allocate a pkt from shm tx pkt pool for every `fi_sendmsg` it posted to shm, and it will return an `EAGAIN` and retry later if the tx pool is drained. I guess we may need to introduce a retry here as well. Let me think about it.